shortfin/python/shortfin_apps/llm/components/kvcache/page_pool.py

```diff
@@ -0,0 +1,159 @@
+    # 32: number of transformer blocks
+    # 2: one for k + one for v
+    # 16: tokens per page
+    # 8: head count (32 heads, but every 4 heads share the same kv buffer)
+    # 128: hidden dimension
+
+
+class PagePool:
+    """Page table based attention cache.
+
+    While internal to a model, the cache is organized with additional structure
+    per page, outside of the model, it is just a list of pages of a certain
+    element type and number of elements (all inner dims are flattened).
+
+    One page table is allocated per device in a fiber. Currently, this is a
+    dense allocation with committed memory but in the future, we may just
+    allocate the address space and lazily populate it with committed memory.
+
+    The cache is unique because usage of it can span fibers and concurrency
+    is implicitly managed at the block level (i.e. freshly acquired blocks
+    are assumed to be uninitialized and available immediately for use).
+
+    It is initialized with a discrete list of fiberd devices from a fiber but
+    cache usage can be done from any fiber which includes those devices.
+
+    In addition to supporting paged attention standalone, this also serves
+    as the array / buffer allocation layer for radix attention described in
+    `radix_tree.py`.
+    """
+
+    def __init__(self, *, devices: Sequence[sf.ScopedDevice], config: PagePoolConfig):
+        self._lock = threading.Lock()
+        self.devices = list(devices)
+        self.config = config
+        self.page_tables: list[sf.array.device_array] = []
+
+        # Setup accounting structs.
+        self.attn_page_entries = [
+            PageInfo(
+                index=i,
+                pool=self,
+                token_offset=0,
+                token_count=0,
+            )
+            for i in range(self.config.alloc_page_count)
+        ]
+
+        self.attn_page_free = list(self.attn_page_entries)
+
+        # Initialize a page table on each device.
+        page_table_shape = [
+            self.config.alloc_page_count,
+            self.config.paged_kv_block_size_elements,
+        ]
+        for device in devices:
+            logging.info(
+                "Allocating page table (shape=%r, dtype=%r, size=%s) on %r",
+                page_table_shape,
+                self.config.dtype,
+                human_size(config.dtype.compute_dense_nd_size(page_table_shape)),
+                device,
+            )
+            page_table = sf.array.device_array.for_device(
+                device, page_table_shape, self.config.dtype
+            )
+            self.page_tables.append(page_table)
+
+    def acquire_free_pages(self, count: int) -> list[PageInfo] | None:
+        with self._lock:
+            available = len(self.attn_page_free)
+            if count > available:
+                return None
+            return [self.attn_page_free.pop() for _ in range(count)]
+
+    def release_pages(self, pages: list[PageInfo]):
```

Comment:
Suppose we have 10 pages that we call `release_pages` on, but 3 of them have non-zero reference counts. What happens in that scenario? Or is that handled already by the `lock`?

Reply:
By the time PagePool.release_pages is called, all the pages should have reference count zero.

I'm still struggling a little with how to do reference counting though. There's the RadixNode (in the RadixTree) which could (and probably should) have a reference count. Then there are the kvcache page which should maytbe also be reference counted. Depending on whether PageInfo->RadixNode is a one-to-one or one-to-many relationship.

Reply:
I could see one-to-many being possible.

Check out example below, and lemme know if you agree that it's a valid case. I could see scenarios where the same node is repeated on a different branch on the radix tree. IIUC, in this scenario, nodes C and F should be pointing to the same KVCache tensors, right?

```mermaid
graph LR
    A[A - SYSTEM PROMPT]
    B[B - User: Hello!<br>Assistant: Hi!]
    C[C - User: What's your favorite color?]
    D[D - User: How are you?]
    E[E - Assistant: Good!]
    F[F - User: What's your favorite color?]

    A --> B
    B --> C
    B --> D
    D --> E
    E --> F

    C -.-> G[...]
    F -.-> H[...]
```

Reply:
Not for most models!

For most language models, in the sentences

He stalks his prey
The wheat stalks dry in the summer sun

"stalks" would have different KV.

There are some models that I vaguely remember would be able to share KV if a stretch of preceeding tokens are the same. I think they're called windowed models? Not sure.